Start using mypy for static type checking

sechkova opened the pull request:
<!-- Please fill in the fields below to submit a pull request.  The more information
that is provided, the better. -->

<!-- Insert issue number here. See https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword for more info. -->
Fixes: #357
Partially #358

### Description of the changes being introduced by the pull request:
Use `mypy` for static type checking. 
- Start with adding type annotations to `util.py` 
- Add initial configuration file `mypy.ini `(needed because `securesystemslib` does not use `setup.cfg` file)
- Add one additional `tox` environment for `mypy` and the corresponding CI workflow

Inspired by https://github.com/theupdateframework/tuf/pull/1395
Developers can run `mypy` directly from the command line or by using `tox -e mypy`.

### Please verify and check that the pull request fulfils the following requirements:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

jku commented on securesystemslib/util.py:
```diff
@@ -246,7 +253,8 @@ def ensure_parent_dir(filename, storage_backend=None):
 
 
-def file_in_confined_directories(filepath, confined_directories):
+def file_in_confined_directories(filepath: str,
+    confined_directories: Union[List[str], Tuple[str]]) -> bool:
```
The tuple case seems weird and completely unused ... but I think your annotation does not work as it defines a tuple of specifically one item.

I think this covers both cases: `confined_directories: Sequence[str]` -- docs say _Tuples implement all of the common sequence operations._ and that should be good enough for us since we only need `x in y`

Could update the method documentation as well.

sechkova replied:
Seems like the correct annotation was `Tuple[str, ...]` but I agree that `Sequence` works just as well and makes things look cleaner.

jku reviewed (changes requested):
I think util.py should `import json` like normal people. 

If someone feels strongly about preserving the weird import_json(), the implementation of that then becomes a simple `return json`

jku commented on securesystemslib/util.py:
```diff
@@ -23,13 +23,16 @@
 
-def get_file_details(filepath, hash_algorithms=['sha256'],
-    storage_backend=None):
+def get_file_details(filepath: str, hash_algorithms: List[str] =['sha256'],
+    storage_backend: Optional[StorageBackendInterface] = None
+    ) -> Tuple[int, Dict[str, str]]:
```
maybe even move filepath and hash_algorithms to new line? :shrug:

sechkova commented:
Edited all three commits with the proposed changes:

- reformatted the function arguments in sort-of-black-style
- replaced Union[List, Tuple] with Sequence
- added mypy to the list of tox environments
- Dropped `import_json` and its test case

jku reviewed (approved):
thanks!

joshuagl commented on securesystemslib/util.py:
```diff
@@ -17,19 +17,25 @@
-    storage_backend=None):
+def get_file_details(
+    filepath: str,
+    hash_algorithms: List[str] =['sha256'],
```
```suggestion
    hash_algorithms: List[str] = ['sha256'],
```

joshuagl commented on securesystemslib/util.py:
```diff
@@ -129,7 +138,10 @@ def get_file_hashes(filepath, hash_algorithms=['sha256'],
-def get_file_length(filepath, storage_backend=None):
+def get_file_length(
+    filepath: str,
+    storage_backend: Optional[StorageBackendInterface]=None
```
```suggestion
    storage_backend: Optional[StorageBackendInterface] = None
```
?

joshuagl commented on securesystemslib/util.py:
```diff
@@ -73,8 +79,11 @@ def get_file_details(filepath, hash_algorithms=['sha256'],
-    storage_backend=None):
+def get_file_hashes(
+    filepath: str,
+    hash_algorithms: List[str] =['sha256'],
```
```suggestion
    hash_algorithms: List[str] = ['sha256'],
```

joshuagl commented on securesystemslib/util.py:
```diff
@@ -160,8 +172,12 @@ def get_file_length(filepath, storage_backend=None):
+def persist_temp_file(
+    temp_file: IO,
+    persist_path: str,
+    storage_backend: Optional[StorageBackendInterface]=None,
```
```suggestion
    storage_backend: Optional[StorageBackendInterface] = None,
```

joshuagl commented on securesystemslib/util.py:
```diff
@@ -160,8 +172,12 @@ def get_file_length(filepath, storage_backend=None):
+    temp_file: IO,
+    persist_path: str,
+    storage_backend: Optional[StorageBackendInterface]=None,
+    should_close: bool=True
```
```suggestion
    should_close: bool = True
```

joshuagl commented on securesystemslib/util.py:
```diff
@@ -200,7 +216,10 @@ def persist_temp_file(temp_file, persist_path, storage_backend=None,
-def ensure_parent_dir(filename, storage_backend=None):
+def ensure_parent_dir(
+    filename: str,
+    storage_backend: Optional[StorageBackendInterface]=None
```
```suggestion
    storage_backend: Optional[StorageBackendInterface] = None
```

joshuagl commented on securesystemslib/util.py:
```diff
@@ -340,8 +361,9 @@ def import_json():
 
 json = import_json()
 
+json_str = Union[str, bytes, bytearray]
```
I think we just need `Union[str, bytes]` here? https://github.com/python/typeshed/blob/master/stdlib/json/__init__.pyi#L36

That should be short enough that we can include it inline?

sechkova replied:
[json's docs](https://docs.python.org/3/library/json.html#json.loads) said that but I agree on the shorter and practical annotation.

joshuagl commented on securesystemslib/util.py:
```diff
@@ -378,7 +356,10 @@ def load_json_string(data):
-def load_json_file(filepath, storage_backend=None):
+def load_json_file(
+    filepath: str,
+    storage_backend: Optional[StorageBackendInterface]=None
```
```suggestion
    storage_backend: Optional[StorageBackendInterface] = None
```

joshuagl commented on mypy.ini:
```diff
@@ -0,0 +1,6 @@
+files = securesystemslib/util.py
+# Supress error messages until enough modules
+# are type annotated
+follow_imports = silent
```
```suggestion
follow_imports = silent

```
Very much a nit, but tools will keep complaining about not having a newline at the end of the file.

joshuagl reviewed (commented):
Thanks for this. I have only nit-picking comments to make, except for one on type annotations.

joshuagl reviewed (approved):
Thanks @sechkova !
